File: 2021/Day-01/task.rb
require '../../aoc_utils'

# solution 1
sum = 0
input_lines
  .map(&:to_i)
  .each_with_previous do |c, p|
    sum += 1 if !p.nil? && c > p
  end

puts sum

# solution 2
sum = 0
prev = nil
input_lines
  .map(&:to_i)
  .each_cons(3) do |elements|
    sum += 1 if prev.not_nil? && elements.sum > prev
    prev = elements.sum
  end

puts sum

File: 2021/Day-01/task.rs
use std::fs;

fn solution1(input: &String) {
    let mut times_increased = 0;
    let mut prev = 0;
    for line in input.lines() {
        let number = line.parse::<i32>().unwrap();
        if number > prev && prev != 0 {
            times_increased += 1;
        }
        prev = number;
    }
    println!("Solution 1: {}", times_increased);
}

fn solution2(input: &String) {
    let mut times_increased = 0;
    let mut prev_sum = 0;
    let parsed_numbers = input.lines()
        .map(|s| s.parse::<i32>().unwrap())
        .collect::<Vec<i32>>();

    for elements in parsed_numbers.windows(3) {
        if prev_sum > 0 {
            if elements.iter().sum::<i32>() > prev_sum {
                times_increased += 1;
            }
        }
        prev_sum = elements.iter().sum::<i32>();
    }
    println!("Solution 2: {}", times_increased);
}

fn main() {
    let input = fs::read_to_string("input1.txt").unwrap();
    solution1(&input);
    solution2(&input);
}